### PR TITLE
Param net

### PR DESCRIPTION
**Pull request by PhMueller**

### CODE REVIEW

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        logger.info(f'Start Benchmark on dataset {dataset}')
+
+        dm = ParamNetDataManager(dataset=dataset)
+        self.surrogate_objective, self.surrogate_costs = dm.load()
+
+    @staticmethod
+    def convert_config_to_array(configuration: Dict) -> np.ndarray:
+        cfg_array = np.zeros(8)
+        cfg_array[0] = 10 ** configuration['initial_lr_log10']
+        cfg_array[1] = round(2 ** configuration['batch_size_log2'])  # todo: round
```
Comment (KEggensperger) [inline hpobench/benchmarks/surrogates/paramnet_benchmark.py]: This todo is outdated?

Comment (KEggensperger) [reply hpobench/benchmarks/surrogates/paramnet_benchmark.py]: Maybe make a comment saying that in contrast to previous versions we now round

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+            Fixing the seed for the ConfigSpace.ConfigurationSpace
+
+        Returns
+        -------
+        ConfigSpace.ConfigurationSpace
+        """
+
+        seed = seed if seed is not None else np.random.randint(1, 100000)
+        cs = CS.ConfigurationSpace(seed=seed)
+        # TODO: in origin benchmark: all uniform float and default is the mid value.
```
Comment (KEggensperger) [inline hpobench/benchmarks/surrogates/paramnet_benchmark.py]: Then why did you change the default value for this version?

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+    @staticmethod
+    def get_fidelity_space(seed: Union[int, None] = None) -> CS.ConfigurationSpace:
+        """
+        Creates a ConfigSpace.ConfigurationSpace containing all fidelity parameters for
+        the SupportVector Benchmark
+
+        Fidelities
+        ----------
+        step: float - [1, 50]
```
Comment (KEggensperger) [inline hpobench/benchmarks/surrogates/paramnet_benchmark.py]: integer

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+                           rng: Union[np.random.RandomState, int, None] = None, **kwargs) -> Dict:
+
+        config_array = self.convert_config_to_array(configuration)
+        lc = self.surrogate_objective.predict(config_array)[0]
+        costs = self.surrogate_costs.predict(config_array)[0]
+
+        # If we can't afford a single epoch, return TODO.
+        if (costs / self.n_epochs) > fidelity['budget']:
+            # TODO: Return random performance here instead
+            y = 1
```
Comment (KEggensperger) [inline hpobench/benchmarks/surrogates/paramnet_benchmark.py]: No, 1 is fine. Are all of these classification problems and we minimize error in [0, 1]?

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+            return {'function_value': float(y),
+                    'cost': fidelity['budget'],
+                    'info': {'state': 'Not enough budget'}}
+
+        learning_curves_cost = np.linspace(costs / self.n_epochs, costs, self.n_epochs)
+
+        if fidelity['budget'] < costs:
+            idx = np.where(learning_curves_cost < fidelity['budget'])[0][-1]
```
Comment (KEggensperger) [inline hpobench/benchmarks/surrogates/paramnet_benchmark.py]: It is very unlikely, but learning_curve_cost==budget can happen. Shouldn't this be <= then?

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+            y = lc[idx]
+            lc = lc[:idx]
+
+        else:
+            # If the budget is larger than the actual runtime, we extrapolate the learning curve
```
Comment (KEggensperger) [inline hpobench/benchmarks/surrogates/paramnet_benchmark.py:222]: This has been done before, could you please add a line to the meta-info where you took the this from (I assume hpolib1.5)

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        Parameters
+        ----------
+        seed : int, None
+            Fixing the seed for the ConfigSpace.ConfigurationSpace
+
+        Returns
+        -------
+        ConfigSpace.ConfigurationSpace
+        """
+        budgets = {  # .   (min, max)-budget (seconds) for the different data sets
```
Comment (KEggensperger) [inline hpobench/benchmarks/surrogates/paramnet_benchmark.py:264]: Please add a link to the HPBandster Code to know where these numbers come from

Comment (KEggensperger) [reply hpobench/benchmarks/surrogates/paramnet_benchmark.py:264]: Ah, okay, if this is also described in the Paper (as in meta-info), then we don't need a link here

File: hpobench/container/recipes/surrogates/Singularity.ParamnetBenchmark [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+VERSION v0.0.1
+
+%post
+    apt update -y \
+    && apt install build-essential git wget -y
+
+    cd /home \
+    && git clone https://github.com/automl/HPOBench.git \
+    && cd HPOBench \
+    && pip install --upgrade .[paramnet] \
```
Comment (KEggensperger) [inline hpobench/container/recipes/surrogates/Singularity.ParamnetBenchmark:15]: Why --upgrade? We don't do this for other benchmarks?

File: hpobench/util/data_manager.py [hunk 9]
```diff
@@ -537,7 +539,6 @@ def __init__(self, data_path: Union[str, Path]):
                             lock_path=f'{hpobench.config_file.cache_dir}/lock_nasbench_101_data', delay=0.5)
     def _download(self):
         from tqdm import tqdm
```
Comment (KEggensperger) [inline hpobench/util/data_manager.py:541]: I am fine with adding tqdm as a dependency, but then it should be part of requirements.txt and not only required for paramnet?

**Review by KEggensperger — CHANGES_REQUESTED** [review]
I left some comments, but overall this looks good to me. Thanks

File: hpobench/benchmarks/surrogates/paramnet_benchmark.py (hunk 2 above)
Comment (PhMueller) [reply hpobench/benchmarks/surrogates/paramnet_benchmark.py]: good question. I've changed the values to the original default values but wrote them explicitly in the constructor so that it is clear what the default values are.

File: hpobench/container/recipes/surrogates/Singularity.ParamnetBenchmark (hunk 8 above)
Comment (PhMueller) [reply hpobench/container/recipes/surrogates/Singularity.ParamnetBenchmark:15]: This is because we need an older scikit-learn version. Saving and restoring an TreeRegressor is only allowed until 0.25 I guess. There is a warning about it. Therefore, I put the new scikit-learn version in the requirements and to make sure it is installed, I've added the --upgrade parameter. 
Without it, pip is already happy if there is it already installed.

File: hpobench/util/data_manager.py (hunk 9 above)
Comment (PhMueller) [reply hpobench/util/data_manager.py:541]: hm, the thing is, we only use it for the paramnet benchmark (more precisely in the paramnet dataloader). The Hpobench package does not rely on it. But if you prefer it more in the general requirements, that's no problem at all :-)

File: hpobench/container/recipes/surrogates/Singularity.ParamnetBenchmark (hunk 8 above)
Comment (KEggensperger) [reply hpobench/container/recipes/surrogates/Singularity.ParamnetBenchmark:15]: Okay, so this is fine then.

File: hpobench/util/data_manager.py (hunk 9 above)
Comment (KEggensperger) [reply hpobench/util/data_manager.py:541]: But this is not the paramnet dataloader, this is NASBench_101. The paramnet dataloader does not make use of tqdm. Okay, but then, since it is only one dataloader, I guess it's best to move the dependency to the respective benchmark requirements.

**Review by KEggensperger — APPROVED** [review]

File: hpobench/util/data_manager.py (hunk 9 above)
Comment (PhMueller) [reply hpobench/util/data_manager.py:541]: ouh, that's a point. I forgot the tqdm requirement for the nasbench101 benchmark. But the paramnet data loader also uses tqdm. (line 628)
